examples/2d/text_input_2d.rs

```diff
@@ -0,0 +1,465 @@
+    },
+    input::keyboard::{Key, KeyboardInput},
+    prelude::*,
+    render::{sync_world::TemporaryRenderEntity, Extract, RenderApp},
+    sprite::Anchor,
+    sprite_render::{
+        ExtractedSlice, ExtractedSlices, ExtractedSprite, ExtractedSpriteKind, ExtractedSprites,
+    },
+    text::{
+        CursorBlink, LineBreak, Motion, Placeholder, PlaceholderLayout, PositionedGlyph,
+        TextBounds, TextCursorBlinkInterval, TextEdit, TextEdits, TextInputBuffer,
+        TextInputSystems, TextInputTarget, TextLayoutInfo,
+    },
+    window::PrimaryWindow,
+};
+
+#[derive(Component)]
+struct TextInputSize(Vec2);
+
+#[derive(Component, Default)]
+struct Overwrite(bool);
+
+fn main() {
+    let mut app = App::new();
+
+    app.add_plugins(DefaultPlugins)
+        .add_systems(Startup, setup)
+        .add_systems(
+            PostUpdate,
+            (update_inputs, update_targets).before(TextInputSystems),
+        )
+        .add_event::<TextSubmission>()
+        .add_systems(Update, handle_submissions);
+    app.sub_app_mut(RenderApp)
+        .add_systems(ExtractSchedule, extract_text_input);
+
+    app.run();
+}
+
+fn setup(mut commands: Commands) {
+    commands.spawn(Camera2d);
+
+    commands.spawn((
+        Text2d::new("submit with SHIFT + ENTER"),
+        Anchor::TOP_CENTER,
+        TextBounds {
+            width: Some(500.),
+            height: None,
+        },
+        TextLayout {
+            linebreak: LineBreak::AnyCharacter,
+            justify: Justify::Left,
+        },
+        Transform::from_translation(Vec3::new(0., -25., 0.)),
+    ));
+
+    commands.spawn((
+        TextInputBuffer {
+            ..Default::default()
+        },
+        CursorBlink {
+            cursor_blink_timer: 0.,
+        },
+        Overwrite::default(),
+        TextInputSize(Vec2::new(500., 250.)),
+        Transform::from_translation(Vec3::new(0., 150., 0.)), // This is in UI coordinates
+        Placeholder::new("type here.."),
+        Visibility::default(),
+        VisibilityClass([TypeId::of::<Sprite>()].into()), // Text input is rendered as sprites
+        Anchor::CENTER,
+    ));
+}
+
+// Here we set the text rendering target's size to match based on the window's scale factor
+fn update_targets(
+    mut commands: Commands,
+    windows: Query<&Window, With<PrimaryWindow>>,
+    mut query: Query<(
+        Entity,
+        &TextInputSize,
+        &mut TextInputTarget,
+        &Anchor,
+        Option<&mut Aabb>,
+    )>,
+) {
+    let scale_factor = windows
+        .single()
+        .map(|window| window.resolution.scale_factor())
+        .unwrap_or(1.0);
+    for (entity, size, mut target, anchor, aabb) in query.iter_mut() {
+        if target.set_if_neq(TextInputTarget {
+            size: size.0 * scale_factor,
+            scale_factor,
+        }) {
+            let (x1, x2, y1, y2) = anchor.reposition(size.0);
+            let new_aabb = Aabb::from_min_max(Vec3::new(x1, y1, 0.), Vec3::new(x2, y2, 0.));
```

Comment:
Specifically this Aabb, and it should be called something like calc bounds or get bounds for.